### PR TITLE
Add pre-connection partner coupon JITM

### PR DESCRIPTION
**Pull request by kallehauge**
### Prerequisite

* https://github.com/Automattic/jetpack/pull/21813 - This PR depends on the coupon logic introduced here.

### Changes proposed in this Pull Request:

* Add a "Partner Pre-connection JITM" that is intended to inform non-connected Jetpack users that they have a "Hosting offer" waiting for them.

The pre-connection JITM system works in a queue, so we simply append the "Coupon Partner"  message in front and if it has been closed, the next JITM will take its place.

### Jetpack product discussion

* PT: pbtFPC-2c3-p2
* Design: p1HpG7-dvB-p2

### Does this pull request change what data or activity we track or use?

* No. This PR only touches "pre connection code" which automatically means: _no tracking_.

### Testing instructions:

* Spin up local version of this branch.
* Add `define( 'SCRIPT_DEBUG', true );` to `wp-config.php`. This is required for the JITM script to work since the minified versions doesn't exist before building the actual plugin.
* Make sure Jetpack is installed (but not connected).
* Add the _**Test partner snippet**_ snippet below to your WordPress installation (e.g.: in 0-sandbox.php if you use Jetpack Docker). This will allow the test coupon to be recognised as a valid coupon.
* Go to `/wp-admin/admin.php?jetpack-partner-coupon=JPTST_JPTA_123AB` (but don't connect). This will just store the coupon so we trigger the pre-connection JITM.
* Go to `/wp-admin/upload.php`.
* You should now see the Coupon JITM (see "Pre-connection partner coupon JITM" screenshot below).
  * If you click the "Redeem Coupon" button, you should be taken to the connection screen that asks up you setup Jetpack and claim a partner product.
  * If you close the JITM (press "X") and refresh the page, you should see another JITM. On the uploads page it's a promotion for lightning fast images (see "Other JITM" screenshot below).

_**Test partner snippet**_
```
add_filter( 'jetpack_partner_coupon_supported_partners', function ( $partners ) {
	$partners['JPTST'] = 'Jetpack Test';

	return $partners;
} );

add_filter( 'jetpack_partner_coupon_supported_presets', function ( $presets ) {
	$presets['JPTA'] = 'jetpack_backup_daily';

	return $presets;
} );
```

### Screenshots

**Pre-connection partner coupon JITM**

Screenshot 2021-11-23 at 18 20 40

**Other JITM**

Screenshot 2021-11-23 at 18 15 28

### CODE REVIEW

**Comment by rcoll** [comment]
This _seems fine_ to me, but I'm having difficulty with a couple things.

1. I haven't been able to get it to actually work. I'm not sure what I'm doing wrong, but I've carefully followed the instructions and nothing appears when I go to the media page. Perhaps because the site was previously connected? I've purged all options and it makes no difference. Yes, I've included the filters. Yes, I've used the correct test coupon code.
2. Why do we need the filter at all? I see what you're trying to do there, but this is an extra step we'd have to ask hosts to follow, which seems suboptimal. Shouldn't we make it easy on them, and just include this code in the plugin?

Code-wise, I'm fine approving this, but under the circumstances I'd feel better seeing it work before signing off.

**Review by rcoll — CHANGES_REQUESTED** [review]
This seems fine to me, but I'm having difficulty with a couple things.

I haven't been able to get it to actually work. I'm not sure what I'm doing wrong, but I've carefully followed the instructions and nothing appears when I go to the media page. Perhaps because the site was previously connected? I've purged all options and it makes no difference. Yes, I've included the filters. Yes, I've used the correct test coupon code.
Why do we need the filter at all? I see what you're trying to do there, but this is an extra step we'd have to ask hosts to follow, which seems suboptimal. Shouldn't we make it easy on them, and just include this code in the plugin?
Code-wise, I'm fine approving this, but under the circumstances I'd feel better seeing it work before signing off.

**Comment by kallehauge** [comment]
> 1. I haven't been able to get it to actually work.

I might have forgotten this step, but if you get a 404 error in console about a JITM script, it can be fixed by enabling script debugging (´define( 'SCRIPT_DEBUG', true );`). It will try and find the "not-minified" version which exists in the monorepo.

> 2. Why do we need the filter at all? I see what you're trying to do there, but this is an extra step we'd have to ask hosts to follow, which seems suboptimal. Shouldn't we make it easy on them, and just include this code in the plugin?

It's not necessarily because we need the filters for partner in the future, they should be added like we do with IONOS; but it does give us the option for hosts to add some sort of backwards compatibility for partner coupons.
In this particular instance, it's so the plugin recognises our test coupon as a real partner coupon and we don't trigger IONOS analytics events during tests etc.

`Jetpack_Partner_Coupon_Helper::get_coupon()` will return false if we don't recognise the partner or preset because we don't know which partner / product to promote in both the JITM and redeem component

**Review by rcoll — APPROVED** [review]
The `SCRIPT_DEBUG` trick worked. Thanks for the explanation on the filters. All is good now.

File: projects/plugins/jetpack/class-jetpack-pre-connection-jitms.php [hunk 1]
```diff
@@ -118,7 +150,7 @@ private function generate_admin_url( $args ) {
 	 * @return array The array of pre-connection JITMs.
 	 */
 	public function add_pre_connection_jitms( $pre_connection_messages ) {
-		$jetpack_messages = $this->get_raw_messages();
+		$jetpack_messages = array_merge( $this->get_raw_messages(), $this->maybe_get_raw_partnership_messages() );
```
Comment (jeherve) [inline projects/plugins/jetpack/class-jetpack-pre-connection-jitms.php:153]: Would it be an option to add the message directly into `get_raw_messages()`? Maybe that would make it easier for folks looking to quickly see the whole list of pre-connection JITMs?

**Review by jeherve — COMMENTED** [review]
This looks good. I only have one question, but not a blocker.

File: projects/plugins/jetpack/class-jetpack-pre-connection-jitms.php (hunk 1 above)
Comment (kallehauge) [reply projects/plugins/jetpack/class-jetpack-pre-connection-jitms.php:153]: We could definitely add them in the same location. My main thought for the separation is that "get_raw_messages" would have to contain partner conditions logic as well ("Should we add these extra messages?") and separation of concern became a thing.

I don't mind updating it if you lean just 1% towards this living in the same method instead 😄

Comment (jeherve) [reply projects/plugins/jetpack/class-jetpack-pre-connection-jitms.php:153]: > I don't mind updating it if you lean just 1% towards this living in the same method instead

Nah, that's alright as it is, it's not that big of a difference. Let's 🚢 this :)

**Pull request merged**